File: 024 - Lexicographic permutations/C++/pe_024.cc
/*
 * file: pe_024.cc
 * title: Lexicographic permutations
 * author: Alexander Gosselin
 * date: September 10, 2013
 * e-mail: alexandergosselin@gmail.com
 *
 */

#include <algorithm>
#include <iostream>
#include <vector>
using namespace std;

int main(int argc, char **argv)
{
  size_t number_of_permutations = 1000000; //print the Nth permutation
  //initialize string of characters to permute
  string characters = "0123456789";
  number_of_permutations--; //first permutation has just been completed
  const size_t number_of_characters = characters.size();
  string permutation; //initialize string containing permutation of characters
  //create array containing factorial(size of array - index)
  vector<unsigned int> factorials;
  factorials.push_back(1);
  for(size_t index = 2; index < number_of_characters; index++) {
    factorials.insert(factorials.begin(), index*factorials[0]);
  }
  /*//return error if N is too large
  if(number_of_permutations > number_of_characters*factorials[0]) {
    cout << "Error: The number of permutations requested is larger than the"
            "total number of possible permutaions.";
    return 0;
  }*/
  //generate Nth permutation of the string
  for(size_t index = 0; index < number_of_characters-1; index++) {
    size_t index_of_target_character = number_of_permutations/factorials[index];
    permutation.push_back(characters[index_of_target_character]);
    characters.erase(characters.begin()+index_of_target_character);
    number_of_permutations -= factorials[index]*index_of_target_character;
  }
  cout << permutation << characters << endl;
  return 0;
}

File: 024 - Lexicographic permutations/Haskell/pe_024.hs
{-
file:   pe_024.hs
title:  Lexicographic permutations
author: Alexander P. Gosselin
e-mail: alexandergosselin@gmail.com
date:   May 25, 2015

link:   https://projecteuler.net/problem=24

copyright:  (C) 2015 Alexander Gosselin
license:    GNU General Public License <http://www.gnu.org/licenses/>
-}

-- Memoized factorial function    
factorial :: Int -> Int
factorial = (map factorial' [0..] !!)
  where
    factorial' 0 = 1
    factorial' n = n * factorial (n - 1)

-- Permute a finite list n times
permutation :: Eq a => Int -> [a] -> [a]
permutation 0 xs = xs -- 1st permutation in Project Euler convention
permutation n xs = x:permutation n' xs'
  where
    x   = xs !! (quot n f)
    n'  = mod n f
    xs' = filter (/= x) xs
    f   = factorial (length xs - 1)

-- The millionth permutation of a list is permutation number 999999
pe_024 = read (permutation 999999 ['0'..'9']) :: Int

main :: IO ()
main = do
    print pe_024

